core/test/solver/batch_idr.cpp

```diff
@@ -0,0 +1,287 @@
+{
+    ASSERT_EQ(this->batchidr_factory->get_executor(), this->exec);
+}
+
+
+TYPED_TEST(BatchIdr, FactoryCreatesCorrectSolver)
+{
+    using Solver = typename TestFixture::Solver;
+    for (size_t i = 0; i < this->nbatch; i++) {
+        ASSERT_EQ(this->solver->get_size().at(i),
+                  gko::dim<2>(this->nrows, this->nrows));
+    }
+    auto batchidr_solver = static_cast<Solver *>(this->solver.get());
+    ASSERT_NE(batchidr_solver->get_system_matrix(), nullptr);
+    ASSERT_EQ(batchidr_solver->get_system_matrix(), this->mtx);
+}
+
+
+TYPED_TEST(BatchIdr, CanBeCopied)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using Solver = typename TestFixture::Solver;
+    auto copy = this->batchidr_factory->generate(Mtx::create(this->exec));
+
+    copy->copy_from(this->solver.get());
+
+    for (size_t i = 0; i < this->nbatch; i++) {
+        ASSERT_EQ(copy->get_size().at(i),
+                  gko::dim<2>(this->nrows, this->nrows));
+    }
+    auto copy_mtx = static_cast<Solver *>(copy.get())->get_system_matrix();
+    const auto copy_batch_mtx = static_cast<const Mtx *>(copy_mtx.get());
+    GKO_ASSERT_BATCH_MTX_NEAR(this->mtx.get(), copy_batch_mtx, 0.0);
+}
+
+
+TYPED_TEST(BatchIdr, CanBeMoved)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using Solver = typename TestFixture::Solver;
+    auto copy = this->batchidr_factory->generate(Mtx::create(this->exec));
+
+    copy->copy_from(std::move(this->solver));
+
+    for (size_t i = 0; i < this->nbatch; i++) {
+        ASSERT_EQ(copy->get_size().at(i),
+                  gko::dim<2>(this->nrows, this->nrows));
+    }
+    auto copy_mtx = static_cast<Solver *>(copy.get())->get_system_matrix();
+    const auto copy_batch_mtx = static_cast<const Mtx *>(copy_mtx.get());
+    GKO_ASSERT_BATCH_MTX_NEAR(this->mtx.get(), copy_batch_mtx, 0.0);
+}
+
+
+TYPED_TEST(BatchIdr, CanBeCloned)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using Solver = typename TestFixture::Solver;
+    auto clone = this->solver->clone();
+
+    for (size_t i = 0; i < this->nbatch; i++) {
+        ASSERT_EQ(clone->get_size().at(i),
+                  gko::dim<2>(this->nrows, this->nrows));
+    }
+    auto clone_mtx = static_cast<Solver *>(clone.get())->get_system_matrix();
+    const auto clone_batch_mtx = static_cast<const Mtx *>(clone_mtx.get());
+    GKO_ASSERT_BATCH_MTX_NEAR(this->mtx.get(), clone_batch_mtx, 0.0);
+}
+
+
+TYPED_TEST(BatchIdr, CanBeCleared)
+{
+    using Solver = typename TestFixture::Solver;
+
+    this->solver->clear();
+
+    ASSERT_EQ(this->solver->get_num_batch_entries(), 0);
+    ASSERT_EQ(this->solver->get_size().get_num_batch_entries(), 0);
+    auto solver_mtx =
+        static_cast<Solver *>(this->solver.get())->get_system_matrix();
+    ASSERT_EQ(solver_mtx, nullptr);
+}
+
+
+TYPED_TEST(BatchIdr, ApplyUsesInitialGuessReturnsTrue)
+{
+    ASSERT_TRUE(this->solver->apply_uses_initial_guess());
+}
+
+
+TYPED_TEST(BatchIdr, CanSetCriteria)
```

Comment:
It would be good to have another similar test where you set the criteria using the setter functions after the solver is generated.

Reply:
I have added such tests now.